### PR TITLE
restructure project with api and messaging

### PR DESCRIPTION
**Pull request by DanielvanBavel**
Update & delete needs a fix

### CODE REVIEW

File: src/infrastructure/mongodb/helpers/mapper.helper.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import { Model } from 'objectmodel';
+import { INewable } from './inewable.interface';
+
+export function mapModelToEntity<TModel extends Document, TEntity>(
+  model: TModel,
+  resultingType: INewable<TEntity>
+) {
+  const obj: any = model.toJSON();
+
+  return new resultingType(obj);
```
Comment (Aplicticus) [inline src/infrastructure/mongodb/helpers/mapper.helper.ts]: Maybe return object instantly? 
return new resultingType(model.toJSON());

File: src/infrastructure/mongodb/model/schema/customer.schema.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  name: {
+    type: String,
+    required: true
+  },
+  email: {
+      type: String,
+      required: true,
+      unique: true
+  },
+  password: {
```
Comment (Aplicticus) [inline src/infrastructure/mongodb/model/schema/customer.schema.ts]: Why do you have password in Customer?

Comment (Aplicticus) [reply src/infrastructure/mongodb/model/schema/customer.schema.ts]: Don't we need N.A.V. information ? (Address, Zipcode etc. )

File: src/infrastructure/mongodb/mongodb.customer.repository.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    if (!updatedModel) {
+      throw new Error('Customer not found');
+    }
+
+    const updatedCustomer = mapModelToEntity<ICustomerDocument, Customer>(updatedModel, Customer);
+
+    return updatedCustomer;
+  }
+
+  public async exists(id: string): Promise<boolean> {
```
Comment (Aplicticus) [inline src/infrastructure/mongodb/mongodb.customer.repository.ts]: Useless method, never used?

File: src/infrastructure/repository/icustomerRepository.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
+import { Customer } from '../../domain/customer';
+
+export interface ICustomerRepository {
+
+    AddCustomer(customer: Customer): Promise<Customer>;
```
Comment (Aplicticus) [inline src/infrastructure/repository/icustomerRepository.ts]: Javascript default code conventions. Please use camelCase.

File: src/infrastructure/repository/icustomerRepository.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+
+    UpdateCustomer(id: string, customer: Customer): Promise<Customer>;
```
Comment (Aplicticus) [inline src/infrastructure/repository/icustomerRepository.ts]: Javascript default code conventions. Please use camelCase.

File: src/infrastructure/repository/icustomerRepository.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+    DeleteCustomer(id: String): Promise<Customer>;
```
Comment (Aplicticus) [inline src/infrastructure/repository/icustomerRepository.ts]: Javascript default code conventions. Please use camelCase.

File: src/infrastructure/services/messagebroker.handler.customer.service.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+        break;
+      }
+      case MessageType.CustomerDeleted: {
+        await this.handleCustomerDeleted(body);
+      }
+    }
+  }
+
+  private async handleCustomerCreated(body: any) {
+    const customer = new Customer(body);
```
Comment (Aplicticus) [inline src/infrastructure/services/messagebroker.handler.customer.service.ts]: Body can be NULL. Maybe add some Nullcheck?

File: src/infrastructure/services/repository.messagebroker.customer.service.ts [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    @inject(TYPES.MessagePublisherProvider)
+    private messagePublisherProvider: MessagePublisherProvider
+  ) {}
+
+  public async add(customer: Customer): Promise<Customer> {
+    const customerRepo = await this.getCustomerRepository();
+
+    // Save it in the repository, since we are sure it is valid now
+    const createdCustomer = await customerRepo.AddCustomer(customer);
+
```
Comment (Aplicticus) [inline src/infrastructure/services/repository.messagebroker.customer.service.ts]: Remove whitespace

File: src/infrastructure/services/repository.messagebroker.customer.service.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+   */
+  private async getCustomerRepository() {
+    return this.customerRepositoryProvider();
+  }
+
+  /**
+   * Gets the message publisher
+   */
+  private async getMessagePublisher() {
+    const t = await this.messagePublisherProvider(RabbitMQExchange.Default);
```
Comment (Aplicticus) [inline src/infrastructure/services/repository.messagebroker.customer.service.ts]: Return the await directly instead to store it in const t.

File: .travis.yml [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    - master
+
+jobs:
+  include:
+    - stage: test
+      language: node_js
+      node_js:
+        - 8
+      cache: yarn
+      script:
```
Comment (Aplicticus) [inline .travis.yml]: Remove Yarn, because you don't have any tests!

File: dockerfile [hunk 11]
```diff
@@ -0,0 +1,38 @@
+FROM node:8-alpine AS build
+
+# Create the app directory
+WORKDIR /app
+
+# Copy package json
+COPY package*.json ./
+
+# COPY yarn.lock
```
Comment (Aplicticus) [inline dockerfile]: Remove Yarn, because you don't have any tests!

Comment (DanielvanBavel) [reply dockerfile]: no because it needed yarn file for the depencies

File: .travis.yml (hunk 10 above)
Comment (DanielvanBavel) [reply .travis.yml]: no

File: dockerfile (hunk 11 above)
Comment (markieo1) [reply dockerfile]: Why is your yarn.lock file empty then??

That means you aren't using yarn

Comment (DanielvanBavel) [reply dockerfile]: removed

**Comment by Aplicticus** [comment]
Test is still existing! 
see: 
https://travis-ci.com/GiethoornInternationalHarbor/CustomerService/builds/74301443?utm_source=github_status&utm_medium=notification

File: dockerfile [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+# Copy package json
+COPY package*.json ./
+
+FROM build as publish
+RUN npm build
+
+FROM node:8-alpine AS runtime
+WORKDIR /app
+
+COPY --from=publish /app/package*.json ./
```
Comment (markieo1) [inline dockerfile]: Copy from dist is also required, since the docker app wont start otherwise. Even though it will build successfully.

**Pull request merged**